activerecord/lib/active_record/attribute_methods.rb

```diff
@@ -165,6 +165,50 @@ def attribute_names
         end
       end
 
+      def enforce_raw_sql_whitelist(args, whitelist: attribute_names_and_aliases) # :nodoc:
+        return if allow_unsafe_raw_sql == :enabled
+
+        unexpected = args.reject do |arg|
+          whitelist.include?(arg.to_s) ||
+            arg.kind_of?(Arel::Node) || arg.is_a?(Arel::Nodes::SqlLiteral)
+        end
+
+        return if unexpected.none?
+
+        if allow_unsafe_raw_sql == :deprecated
+          ActiveSupport::Deprecation.warn(
+            "Dangerous query method (method who'se arguments are used as raw " +
+            "SQL) called with non-attribute argument(s): " +
+            "#{unexpected.map(&:inspect).join(", ")}. Non-attribute " +
+            "arguments will be disallowed in Rails 6.0. This method should " +
+            "not be called with user-provided values, such as request " +
+            "parameters or model attributes. Known-safe values can be passed " +
+            "by wrapping them in Arel.sql()."
+          )
+        else
+          raise(ActiveRecord::UnknownAttributeReference,
+            "Query method called with non-attribute argument(s): " +
+            unexpected.map(&:inspect).join(", ")
+          )
+        end
+      end
+
+      # Can the given name be treated as a column name? Returns true if name
+      # is attribute or attribute alias.
+      #
+      #   class Person < ActiveRecord::Base
+      #   end
+      #
+      #   Person.respond_to_attribute?(:name)
+      #   # => true
+      #
+      #   Person.respond_to_attribute?("foo")
+      #   # => false
+      def respond_to_attribute?(name) # :nodoc:
```

Comment:
This method takes attribute aliases into account, which `attribute_method?` does not.